fix(compiler-cli): report more accurate diagnostic for invalid import

crisbeto opened the pull request:
Currently when an incorrect value is in the `imports` array, we highlight the entire array which can be very noisy for large arrays. This comes up semi-regularly (at least for me) when an import is missing.

These changes add some logic that reports a more accurate diagnostic location for the most common case where the `imports` array is static. Non-static arrays will fall back to the current behavior.

Example of what I'm referring to:
<img width="901" alt="image" src="https://github.com/user-attachments/assets/494208b6-0944-40a2-b0ec-3db62af1df85" />

devversion commented on packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts:
```diff
@@ -132,7 +138,21 @@ export function validateAndFlattenComponentImports(
       );
     } else {
-      diagnostics.push(createValueHasWrongTypeError(expr, imports, errorMessage).toDiagnostic());
+      const isStaticArray =
```
Optional nit: I wonder if it makes sense to just limit to `isArrayExpression(expr) && expr.elements.every(x => ts.isIdentifier(x)) && !imports.some(Array.isArray)`. I don't know for what else this logic might be true.

crisbeto replied:
I was trying to account for this case which I think is allowed:

```
const IMPORTS = [...];

imports: [Foo, IMPORTS];
```

devversion replied:
ah I see.  Aren't you excluding that cass below with the some `isArray` check?

crisbeto replied:
Ah sorry, I thought your snippet excluded the `isArray` check but it's actually the length check. I put it in there as a quicker way to exit in case there's a spread.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts:
```diff
@@ -132,7 +138,21 @@ export function validateAndFlattenComponentImports(
+        ? createValueHasWrongTypeError(expr.elements[i], ref, errorMessage)
+        : createValueHasWrongTypeError(expr, imports, errorMessage);
+
+      diagnostics.push(error.toDiagnostic());
```
For `DynamicValue` we have access to the `ts.Node`, making it possible to use some existing logic:

```ts
      let diagnosticNode = expr;
      if (ref instanceof DynamicValue && ts.isExpression(ref.node)) {
        diagnosticNode = getOriginNodeForDiagnostics(ref.node, expr);
      }
      diagnostics.push(createValueHasWrongTypeError(diagnosticNode, imports, errorMessage).toDiagnostic());
```

but I realise this wouldn't work for e.g. object literals that are evaluated to a `Map`, without access to the originating node. The current approach doesn't have this limitation, but feels oddly specific.

crisbeto replied:
Added some logic for dynamic values, but I left the array-specific one in case the value isn't dynamic, but still isn't valid.

crisbeto committed:
fix(compiler-cli): report more accurate diagnostic for invalid import

Currently when an incorrect value is in the `imports` array, we highlight the entire array which can be very noisy for large arrays. This comes up semi-regularly (at least for me) when an import is missing.

These changes add some logic that reports a more accurate diagnostic location for the most common case where the `imports` array is static. Non-static arrays will fall back to the current behavior.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts:
```diff
@@ -132,7 +139,32 @@ export function validateAndFlattenComponentImports(
+      let diagnosticValue: ResolvedValue;
+
+      if (ref instanceof DynamicValue) {
+        diagnosticNode = ref.node;
```
Without `getOriginNodeForDiagnostics` I think the following may perhaps become unclear:

```ts
const IMPORTS = [HelloDir, InvalidReference];

@Component({ imports: IMPORTS })
export class MyCmp {}
```

Assuming `InvalidReference` is a `DynamicValue` (e.g. because an import is missing/invalid), the diagnostic would be reported within the `IMPORTS` initializer, _separate_ from the component declaration. This becomes more unclear when `IMPORTS` is declared in a different source file.

I suspect that using `getOriginNodeForDiagnostics` would _mostly_ behave the same as the explicit array literal check covers (making it redundant), except for cases where the imports array literal contains a spread element (so possibly still meaningful).

pkozlowski-opensource commented:
This PR was merged into the repository by commit 29eded6457b72e7c8a2d53c847ec43996b73c054.

The changes were merged into the following branches: main, 19.2.x
